File: cliente.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>

#define BUFFER_SIZE 1024

void error_exit(const char *message) {
    perror(message);
    exit(EXIT_FAILURE);
}

int main(int argc, char *argv[]) {
    if (argc != 4) {
        fprintf(stderr, "Uso: %s <IP del servidor> <puerto> <acción>\n", argv[0]);
        fprintf(stderr, "Ejemplo: %s 127.0.0.1 8080 LIST_USERS\n", argv[0]);
        exit(EXIT_FAILURE);
    }

    const char *server_ip = argv[1];
    int server_port = atoi(argv[2]);
    const char *action = argv[3];

    // Crear el socket
    int client_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (client_socket < 0) {
        error_exit("Error al crear el socket");
    }

    // Configurar la dirección del servidor
    struct sockaddr_in server_addr;
    memset(&server_addr, 0, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(server_port);

    if (inet_pton(AF_INET, server_ip, &server_addr.sin_addr) <= 0) {
        error_exit("Dirección IP inválida");
    }

    // Conectar al servidor
    if (connect(client_socket, (struct sockaddr *)&server_addr, sizeof(server_addr)) < 0) {
        error_exit("Error al conectar al servidor");
    }

    printf("Conectado al servidor %s:%d\n", server_ip, server_port);

    // Enviar la acción al servidor
    if (send(client_socket, action, strlen(action) + 1, 0) < 0) {
        error_exit("Error al enviar la acción al servidor");
    }

    // Recibir el código de respuesta del servidor
    char response_code;
    if (recv(client_socket, &response_code, 1, 0) <= 0) {
        error_exit("Error al recibir el código de respuesta del servidor");
    }

    printf("Código de respuesta del servidor: %d\n", response_code);

    // Si el código es 0, recibir la lista de usuarios conectados
    if (response_code == 0) {
        char buffer[BUFFER_SIZE];
        memset(buffer, 0, sizeof(buffer));

        if (recv(client_socket, buffer, sizeof(buffer), 0) <= 0) {
            error_exit("Error al recibir la lista de usuarios del servidor");
        }

        printf("Lista de usuarios conectados:\n%s\n", buffer);
    }

    // Cerrar el socket
    close(client_socket);
    printf("Conexión cerrada.\n");

    return 0;
}

File: cliente.py
import socket
import sys
import argparse

def main():
    # Configurar los argumentos de línea de comandos
    parser = argparse.ArgumentParser(description="Cliente para conectarse al servidor.")
    parser.add_argument("-s", "--server", required=True, help="Dirección IP del servidor")
    parser.add_argument("-p", "--port", required=True, type=int, help="Puerto del servidor")
    args = parser.parse_args()

    server_ip = args.server
    server_port = args.port

    # Crear el socket
    try:
        client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    except socket.error as e:
        print(f"Error al crear el socket: {e}")
        sys.exit(1)

    # Conectar al servidor
    try:
        client_socket.connect((server_ip, server_port))
        print(f"Conectado al servidor {server_ip}:{server_port}")
    except socket.error as e:
        print(f"Error al conectar al servidor: {e}")
        sys.exit(1)

    # Bucle para enviar solicitudes
    try:     
        while True:     
            # Leer la solicitud del usuario
            user_input = input("Ingrese su solicitud: ").strip()
            
            if user_input == "9":
                print("Cerrando conexión...")
                break

            # Validar y procesar la solicitud
            parts = user_input.split(" ")
            if len(parts) < 2:
                print("Formato inválido. Debe incluir una acción y al menos un argumento.")
                continue

            action = parts[0].upper()
            arguments = parts[1:]

            # Construir el mensaje según la acción
            if action in ["REGISTER", "UNREGISTER", "DISCONNECT", "LIST_USERS"]:
                # Estas acciones esperan cadenas separadas por '\0'
                message = f"{action}\0{arguments[0]}\0"
            elif action == "CONNECT":
                # CONNECT espera username y puerto separados por '\0'
                if len(arguments) != 2:
                    print("Formato inválido. CONNECT requiere <username> <port>.")
                    continue
                message = f"{action}\0{arguments[0]}\0{arguments[1]}\0"
            elif action == "PUBLISH":
                # PUBLISH espera username, file_name y description separados por '\0'
                if len(arguments) != 3:
                    print("Formato inválido. PUBLISH requiere <username> <file_name> <description>.")
                    continue
                message = f"{action}\0{arguments[0]}\0{arguments[1]}\0{arguments[2]}\0"
            elif action == "DELETE":
                # DELETE espera username y file_name separados por '\0'
                if len(arguments) != 2:
                    print("Formato inválido. DELETE requiere <username> <file_name>.")
                    continue
                message = f"{action}\0{arguments[0]}\0{arguments[1]}\0"
            elif action == "LIST_CONTENT":
                # LIST_CONTENT espera username y target_username separados por '\0'
                if len(arguments) != 2:
                    print("Formato inválido. LIST_CONTENT requiere <username> <target_username>.")
                    continue
                message = f"{action}\0{arguments[0]}\0{arguments[1]}\0"
            else:
                print("Acción no reconocida.")
                continue

            # Enviar el mensaje al servidor
            try:
                client_socket.sendall(message.encode('utf-8'))
            except socket.error as e:
                print(f"Error al enviar el mensaje al servidor: {e}")
                break

            # Recibir el código de respuesta del servidor
            try:
                response_code = client_socket.recv(1)
                if not response_code:
                    print("El servidor cerró la conexión.")
                    break

                response_code = int.from_bytes(response_code, byteorder='big')
                print(f"Código de respuesta del servidor: {response_code}")

                # Si el código es 0, recibir información adicional (si aplica)
                if response_code == 0:
                    buffer = b""
                    while True:
                        data = client_socket.recv(1024)
                        if not data:
                            break
                        buffer += data
                    if buffer:
                        print("Respuesta del servidor:")
                        print(buffer.decode('utf-8'))

            except socket.error as e:
                print(f"Error al recibir la respuesta del servidor: {e}")
                break

    except KeyboardInterrupt:
        print("\nCerrando conexión...")
    finally:
        client_socket.close()
        print("Conexión cerrada.")

if __name__ == "__main__":
    main()